source/extensions/formatter/metadata/metadata.cc

```diff
@@ -46,6 +46,24 @@ class ListenerMetadataFormatter : public ::Envoy::Formatter::MetadataFormatter {
             }) {}
 };
 
+// Metadata formatter for virtual host metadata.
+class VirtualHostMetadataFormatter : public ::Envoy::Formatter::MetadataFormatter {
+public:
+  VirtualHostMetadataFormatter(const std::string& filter_namespace,
+                               const std::vector<std::string>& path,
+                               absl::optional<size_t> max_length)
+      : ::Envoy::Formatter::MetadataFormatter(
+            filter_namespace, path, max_length,
+            [](const StreamInfo::StreamInfo& stream_info)
+                -> const envoy::config::core::v3::Metadata* {
+              Router::RouteConstSharedPtr route = stream_info.route();
+              if (route == nullptr) {
+                return nullptr;
+              }
+              return &route->routeEntry()->virtualHost().metadata();
```

Comment:
Note, there is no guarantee that the route always has a valid route entry.

Reply:
done handling the invalid route entry https://github.com/envoyproxy/envoy/pull/34958/commits/153fc582e3d70e1d7dbba8b0e82d91c9634cb7a1